api/src/wfl/executor.clj

```diff
@@ -428,6 +446,88 @@
+        ;;    we must resubmit the snapshot in its entirety.
+        ;; 2. Though we are technically resubmitting snapshot references
+        ;;    and not submissions, this query takes a submission-based approach.
+        ;;    Why? Because there is no ambiguity when linking a submission
+        ;;    to its workflows.  With the capacity for retries, one reference
+        ;;    may map to multiple sets of workflows, and retrying the reference
+        ;;    should only update the sibling workflow records of those supplied.
+        query "SELECT * FROM %s
+               WHERE submission IN
+                 (SELECT DISTINCT submission FROM %s
+                 WHERE workflow IN %s)"]
+    (->> (format query details details workflow-ids)
+         (jdbc/query tx)
+         (group-by :reference))))
+
+(defn update-retried-workflow-records
+  "Match `original-records` with `retry-records` on entity uuid,
+  update each original record to point to its retry record,
+  and propagate all updates to `details` table."
+  [{:keys [details] :as executor} [retry-records original-records]]
+  (letfn [(link-retry-to-original
+            [{:keys [entity]         :as original}
+             {:keys [retryEntity id] :as retry}]
+            (when-not (= entity retryEntity)
+              (throw (ex-info (str "Failed to link retry to original workflow: "
+                                   "mismatched entities")
+                              {:original original
+                               :retry    retry
+                               :executor executor})))
+            (assoc original :retry id))
+          (write-retries
+            [now records]
+            (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+              (run!
+               (fn [{:keys [id retry]}]
+                 (jdbc/update! tx details
+                               {:retry retry :updated now}
+                               ["id = ?" id]))
+               records)))]
+    (when-not (== (count retry-records) (count original-records))
+      (throw (ex-info "All original records should be retried"
+                      {:retry-records    retry-records
+                       :original-records original-records
+                       :executor         executor})))
+    (log/debug (format "%s Linking %s retry records to their originals..."
+                       (log-prefix executor) (count retry-records)))
+    (->> retry-records
+         (map #(rename-keys % {:entity :retryEntity}))
+         (sort-by :retryEntity)
+         (map link-retry-to-original (sort-by :entity original-records))
+         (write-retries (utc-now)))))
+
+(defn ^:private retry-terra-executor
+  "Resubmit the snapshot references associated with `workflows` in `workspace`
+  and update each original workflow record with the row ID of its retry."
+  [{:keys [workspace] :as executor} workflows]
+  (letfn [(retried-workflow-records [reference-id]
```

Comment:
`submit-reference`?

Reply:
This one's interesting, another case where I named based on return type.  In my development I didn't realize initially that `allocate-submission` returned the newly allocated records, but I've since updated its docstring.

I think your suggestion works in context of the threading below, since it is passed to `update-retried-workflow-records`.